packages/flutter/lib/src/painting/image_stream.dart

```diff
@@ -196,54 +227,125 @@ abstract class ImageStreamCompleter extends Diagnosticable {
   /// occurred. If the listener is added within a render object paint function,
   /// then use this flag to avoid calling [RenderObject.markNeedsPaint] during
   /// a paint.
-  void addListener(ImageListener listener) {
-    _listeners.add(listener);
-    if (_current != null) {
+  void addListener(ImageListener listener, { ImageErrorListener onError }) {
+    _listeners.add(new _ImageListenerPair(listener, onError));
+    if (_currentImage != null) {
       try {
-        listener(_current, true);
+        listener(_currentImage, true);
       } catch (exception, stack) {
-        _handleImageError('by a synchronously-called image listener', exception, stack);
+        reportError(
+          context: 'by a synchronously-called image listener',
+          exception: exception,
+          stack: stack,
+        );
       }
     }
+    if (_currentError != null && onError != null) {
+      try {
+          onError(_currentError.exception, _currentError.stack);
+        } catch (exception, stack) {
+          FlutterError.reportError(
+            new FlutterErrorDetails(
+              exception: exception,
+              library: 'image resource service',
+              context: 'by a synchronously-called image error listener',
+              stack: stack,
+            ),
+          );
+        }
+    }
   }
 
-  /// Stop listening for new concrete [ImageInfo] objects.
+  /// Stop listening for new concrete [ImageInfo] objects and errors from
+  /// its associated [ImageErrorListener].
   void removeListener(ImageListener listener) {
-    _listeners.remove(listener);
+    _listeners.removeWhere((_ImageListenerPair listenerPair) {
+      return listenerPair.listener == listener;
+    });
   }
 
   /// Calls all the registered listeners to notify them of a new image.
   @protected
   void setImage(ImageInfo image) {
-    _current = image;
+    _currentImage = image;
     if (_listeners.isEmpty)
       return;
-    final List<ImageListener> localListeners = new List<ImageListener>.from(_listeners);
+    final List<ImageListener> localListeners = _listeners.map<ImageListener>(
+      (_ImageListenerPair listenerPair) => listenerPair.listener
+    ).toList();
     for (ImageListener listener in localListeners) {
       try {
         listener(image, false);
       } catch (exception, stack) {
-        _handleImageError('by an image listener', exception, stack);
+        reportError(
+          context: 'by an image listener',
+          exception: exception,
+          stack: stack,
+        );
       }
     }
   }
 
-  void _handleImageError(String context, dynamic exception, dynamic stack) {
-    FlutterError.reportError(new FlutterErrorDetails(
+  /// Calls all the registered error listeners to notify them of an error that
+  /// occurred while resolving the image.
+  ///
+  /// If the same error listener is attached with multiple listeners, that
+  /// error listener will only be notified once.
```

Comment:
why?

Reply:
I even wrote a test to assert otherwise :) Don't know how I forgot to update the docs. In follow-up PR